### PR TITLE
[MRG+1] Float colors for plot_compare_evoked

### LINKED ISSUE #4753: dedicated float option for mne.viz.plot_compare_evokeds
As discussed between @mmagnuski and @agramfort , the color gradient should allow 0-1 float inputs.
I will do it.

See #4526 

closed by #4775

### PR DESCRIPTION
**Pull request by jona-sassenhagen**
@mmagnuski 

Closes #4753

### CODE REVIEW

File: mne/viz/tests/test_evoked.py [hunk 1]
```diff
@@ -217,6 +218,13 @@ def test_plot_evoked():
                     contrasts, colors=colors, picks=[0], cmap='Reds',
                     split_legend=split, linestyles=linestyles,
                     ci=False, show_sensors=False)
+        colors = {"a" + str(ii): ii / len(evokeds)
+                  for ii, _ in enumerate(evokeds)}
+        print(colors)
```
Comment (jona-sassenhagen) [inline mne/viz/tests/test_evoked.py]: I don't get where this is coming from ... it's not on my local branch.

File: mne/viz/evoked.py [hunk 2]
```diff
@@ -1526,18 +1527,38 @@ def _setup_styles(conditions, styles, cmap, colors, linestyles):
...
+                if not 0 <= color <= 1:
+                    raise ValueError("Values of colors must be all-integer or "
+                                     "floats between 0 and 1, got %s." % color)
+            msg = "Float colors detected, mapping to percentiles ..."
+            n_colors = 101  # percentiles plus 1 if we have 1.0s
+            colors_old = colors.copy()
+
+            def convert_colors(color):
+                return int(colors_old[color] * 100)
+            colors_are_float = True
```
Comment (larsoner) [inline mne/viz/evoked.py]: you are already setting `msg` inside these branches, so might as well also set `colors_are_float` in each branch for consistency (i.e., move the `colors_are_float = True` from 1518 up here

Comment (jona-sassenhagen) [reply mne/viz/evoked.py]: I also need it for the `cmap=None` case, 1519.

Comment (larsoner) [reply mne/viz/evoked.py]: ahh okay

**Comment by jona-sassenhagen** [comment]
This is green btw, the error is in #4778

**Comment by agramfort** [comment]
can you document this in example or tuturial? thanks

**Comment by jona-sassenhagen** [comment]
@agramfort example:

https://6528-1301584-gh.circle-artifacts.com/0/tmp/circle-artifacts.qz0Zv56/html/auto_examples/stats/plot_sensor_regression.html

**Comment by agramfort** [comment]
I don't think it's the right place to demo a visualization option. This example is for linear regression raw. Also this seems a bit artificial to me. Do we have a better option/dataset? to set the color an ERP/ERF we could use nave? or ERF/ERP of a word and color by it's frequency?

**Comment by jona-sassenhagen** [comment]
I thought it would be useful to help people who don't immediately get what continuous-predictor regression of EEG does, I think @larsoner asked for this in the tutorial PR ...
This kind of plot is essentially what I'd be using it for - to check for nonlinearities before doing a linear regression on e.g. word frequency or concreteness.

I can move it to the metadata tutorial?

I can also do word frequency instead (although I think the Concreteness example looks pretty good?). We can't really do per-word because there's 960 words.

I don't think we have datasets with continuous predictors. To get some, we could:
- look at something like pre-stimulus activity and bin by that
- RT in the EEGLAB sample data (probably too few trials)
- local probability of audvis/LR sample data

File: examples/stats/plot_sensor_regression.py [hunk 3]
```diff
@@ -46,9 +47,25 @@
...
+    values /= values.max()
+    values = values.round(1)
+    colors = {str(val): val for val in values}
+    df = epochs.metadata
+    name_ = name + "_percentile"
+    df[name_] = values
+    epochs.metadata = df
+    evokeds = dict()
+    for val in epochs.metadata[name_].unique():
+        evokeds[str(val)] = epochs[name_ + " == " + str(val)].average()
```
Comment (agramfort) [inline examples/stats/plot_sensor_regression.py]: I just feel that this code looks too ad hoc with more code to hack the metadata than doing the analysis.

maybe doing it just for Contreteness, without for loop using pd.cut can slim this down to less than 5 lines?

Comment (jona-sassenhagen) [reply examples/stats/plot_sensor_regression.py]: Ok, I see your point. I will try!

**Comment by agramfort** [comment]
example looks good https://6544-1301584-gh.circle-artifacts.com/0/tmp/circle-artifacts.iQnQgAA/html/auto_examples/stats/plot_sensor_regression.html#sphx-glr-auto-examples-stats-plot-sensor-regression-py

go to go on my side

**Comment by larsoner** [comment]
The example looks good to me, so I'm okay with merging as is.

@jona-sassenhagen you want to go the extra mile, though, it could benefit from some narrative and `######`-section separation. Maybe the first section as "you can bin and then plot evokeds, this is what you see" with the binned plot, followed by another section "but if you do regression instead, you see X which has the benefits of Y over the naive binning method" with its plots. I'm trying to think of it as how to sell someone on the method and explain why it's useful, rather than just show them that the methods exist in MNE.

**Comment by jona-sassenhagen** [comment]
Ok I will!

________________________________

**Comment by jona-sassenhagen** [comment]
https://6550-1301584-gh.circle-artifacts.com/0/tmp/circle-artifacts.PEbqYng/html/auto_examples/stats/plot_sensor_regression.html

Looks good IMO!

Feel free to merge.

**Comment by jona-sassenhagen** [comment]
@mmagnuski if you have any comments ...

**Comment by agramfort** [comment]
looks great. thx @jona-sassenhagen

**Comment by mmagnuski** [comment]
I can look in more detail this evening, but just looking at the examples - I love the one with concreteness, and was wondering if the last two figures (Intercept and concreteness) were big enough, they are visibly smaller than the global field power figure (probably because both are created in the same cell).

**Comment by mmagnuski** [comment]
I was a bit too late to the party :)

**Comment by jona-sassenhagen** [comment]
You're right - it would be easy to create a cell for the intercept, and maybe functionally helpful.

**Pull request merged**